src/api/services/ListingItemTemplateService.ts

```diff
@@ -321,59 +321,49 @@ export class ListingItemTemplateService {
+
+        // Determine the highest percentage that we can compress the images to so that they still fit within limits
+        for ( ; qualityFactor <= 1.0; qualityFactor += qualityFactorStep) {
+            let compressedSizeTotal = 0;
+            for (const originalImageData of originalImageDatas) {
+                // Yes, we load the file each time. The alternative is reading and storing the data in memory.
+                //  Which could be used to cause problems with large enough files. So trading execution (I/O) speed for safety
+                const compressedImage = await this.getResizedImage(originalImageData.imageHash, qualityFactor * 100);
+                compressedSizeTotal += compressedImage.length;
+                if (compressedSizeTotal > imageSpaceAvail) {
                     break;
                 }
-
-                this.log.debug('image: ' + itemImage.hash + ', downgrading quality...');
-
-                // need to compress more
-                const evenMoreCompressedImageData = await ImageProcessing.downgradeQuality(
-                    compressedImageData, ListingItemTemplateService.FRACTION_TO_COMPRESS_BY);
-
-                if (compressedImageData.length !== evenMoreCompressedImageData.length) {
-                    // we have not yet reached the limit of compression.
-                    compressedImageData = evenMoreCompressedImageData;
-                    continue;
-                } else {
-                    // sizes equal, so we reached the limit of compression, need to resize the image
-                    numResizings++;
-                    if (numResizings >= ListingItemTemplateService.MAX_RESIZES) {
-                        // a generous number of resizes has happened, but we haven't found a solution yet.
-                        // exit incase this is an infinite loop.
-                        throw new MessageException('After ${numResizings} resizes we still didn\'t compress the image enough.'
-                            + ' Image size = ${compressedImage.length}.');
-                    }
-
-                    this.log.debug('image: ' + itemImage.hash + ', reached the limit of compression, resizing image...');
-
-                    // we've reached the limits of compression. We need to resize the image for further size losses.
-                    compressedImageData = await ImageProcessing.resizeImageToFraction(
-                        originalImageData, ListingItemTemplateService.FRACTION_TO_RESIZE_IMAGE_BY);
-                    originalImageData = compressedImageData;
-                    continue;
-                }
             }
+            if (compressedSizeTotal > imageSpaceAvail) {
+                break;
+            }
+        }
+
+        // If still at lowest quality, then leave quality factor as is. Otherwise, reduce to previous (good) quality setting
+        const actualQualityFactor = qualityFactor === ListingItemTemplateService.FRACTION_LOWEST_COMPRESSION ?
```

Comment:
lol, what

Reply:
Hehehe... yeah, so I included the comment at about 2am :)

The general idea being:
- check if we're actually got beyond the lowest quality setting used;
    - if not, then well, use the lowest value;
    - else, since the current qualityFactor value increases the total image sizes to beyond what is acceptable, decrease the qualityFactor value by the step that was used to increase it. Which means that we're back at a qualityFactor value that allows the messages to fit.

So I was trying to compress the above explanation into that one line :laughing: